include/usb/usb_ch9.h

```diff
@@ -187,12 +187,27 @@ struct usb_if_descriptor {
 	uint8_t iInterface;
 } __packed;
 
+struct usb_ep_desc_bmattr {
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+	uint8_t transfer : 2;
+	uint8_t synch: 2;
+	uint8_t usage: 2;
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+	uint8_t usage : 2;
+	uint8_t synch : 2;
+	uint8_t transfer : 2;
+#endif
+} __packed;
```

Comment:
We should not use bit fields in the part of code that gets built by multiple toolchains targeting multiple architectures, as the memory layout of C/C++ bit fields is implementation-defined and may not necessarily match the order/size defined here.

Reply:
We use it in other subsystems, USB, Bluetooth, 802154. And it has already been discussed. If you still see problems with it then open a general issue. Until then no reason for you to block here.

Reply:
> We use it in other subsystems, USB, Bluetooth, 802154.

We only have very few instances of bit fields being actually used to read/write structured memory data, and that seems to have been an oversight.

> And it has already been discussed.

Where was it discussed? Can you provide a link to this discussion?

Reply:
We have discussed this in the past multiple times, just not sure where exactly. Consensus was that, since the use of the `__BYTE_ORDER__ ` is very extended among firmware developers, and is also used in Linux, that we should continue to benefit from the compiler knowing about the bitfields so it can optimize. If this ever turns out to be a real problem (unlikely since this works fine with GCC and Clang, and there aren't many other frontends anymore out there) then we can either add a new macro that deals with it or fall back to masks.

Reply:
> If this ever turns out to be a real problem (unlikely since this works fine with GCC and Clang, and there aren't many other frontends anymore out there) then we can either add a new macro that deals with it or fall back to masks.

It is true that most compilers follow the order specified in the struct, and there should not be any real issues as long as we make the struct "packed."

However, that does not change the fact that we are at the mercy of the compiler being used, and this may lead to serious portability issues in the future (e.g. this will be a showstopper if we are adding a new toolchain support and that toolchain happens to handle the bit fields in its own implementation-defined way, as permitted by the C standards).

> We have discussed this in the past multiple times, just not sure where exactly. Consensus was that, since the use of the __BYTE_ORDER__ is very extended among firmware developers, and is also used in Linux,

I do not know when that discussion happened or, if any, decision was made; but, if that was a while ago, I would argue that things have changed since then because we now have multiple third-party toolchains (notably Arm Compiler 6 and ARC MWDT) and will likely keep adding new ones in the foreseeable future (e.g. IAR).

One could argue that using non-portable constructs like struct bit fields undermines the efforts made in the toolchain abstraction and portability area.

With that said, the only concern I have at this time is the newly added toolchains such as Arm Compiler 6 and ARC MWDT (cc @tejlmand @evgeniy-paltsev @abrodkin).

Reply:
@stephanosio this is common in other subsystems as @jfischer-no mentions, including Bluetooth:

https://github.com/zephyrproject-rtos/zephyr/blob/f9db7fe0657b7307e48d42aff8a8030eb4ddecc6/subsys/bluetooth/controller/ll_sw/pdu.h#L375-L380

While you are correct in principle, in practice there is no compiler out there that doesn't respect this (that I know of).